### ISSUE TITLE
Deselect dates method does not work in certain rare scenarios

### ISSUE THREAD

**Issue by reden87** [patchthecode/JTAppleCalendar#679]
To reproduce the problem you need to do the following steps:

1) in the **TestViewController.swift** set some start and end date in the near past/future:

```swift
let startDate = formatter.date(from: "2018 01 15")!
let endDate = formatter.date(from: "2018 03 15")!
```

2) in the `didSelectDate` delegate method call the deselectdate method programmatically:

```swift
calendar.deselect(dates: [date])
```

This should programmatically deselect the selected dates instantly (so basically you should NOT be able to select any date at all. Same result would be if you would implement the `shouldSelectDate` method with `return true`).

The bug is that you can select 31 of December and all days between 1 of April and 7 of April.

Basically you are able to select the days before the start date's month and days after the end date's month.

I know this is not you should deselect dates but it's a bug anyways, maybe could cause some problems in other situations.

Hope this description is easier to understand than my previous bug reports.

**Comment by patchthecode** [comment]
I just had time to re-read this.
OK, since this libray's creation I had this:

The startDate always starts at the beginning of the month you choose.
The endDate always ends at the end of the month you choose.

I did it this way becase i did not know how a developer would want the calendar to look.
The problem i had was -> If developer puts start date to be `2018 01 15` then 

1. should the library still generate date cells from from the `1st to the 14th`?
2. or should it leave it as blank spaces? 

So i came up with a better solution for start date generation.  The solution was an `generateInDates` and `generateOutDates` parameter which made the library very flexible. I explained it here in this video -> https://www.youtube.com/watch?v=1Sta3aEo9E4

So basically i should actually rename the parameters you mentioned above to be
```swift
let startMonth = formatter.date(from: "2018 01 15")!
let endMonth = formatter.date(from: "2018 03 15")!
```

And the preventing of a date from being selected should be in the `shouldSelect` function.

Now what if you really DID want your calendar to restrict/stop your users from scrolling to dates before `2018 01 15` or after `2018 03 15` ?  Well dont worry, you can still do so using the [boundaryDates](https://github.com/patchthecode/JTAppleCalendar/issues/365) technique.

Let me know if this helps.

**Comment by wainglaister** [comment]
There appears to be a typo error in the deselect function which means it actually selects instead:

    public func deselect(dates: [Date], triggerSelectionDelegate: Bool = true) {
        if allowsMultipleSelection {
            selectDates(dates, triggerSelectionDelegate: triggerSelectionDelegate)

**Comment by patchthecode** [comment]
@wainglaister you are correct. Fixing this now.
But this doesnt seem to be related to the original poster's bug.
But its a bug  non the less. So i'll fix 👍

**Comment by obeal** [comment]
There's no typo error 
... selectDates function selects or deselects dates depending on whether they're selected or not.
... but, as the original poster's bug reports, there are issues with deselect....

**Comment by patchthecode** [comment]
@obeal i just wish someone would provide me a zipped empty project with just a calendar, and the bug. Then i will fix.

**Comment by obeal** [comment]
OK.  @patchthecode 

From what I've seen the _selectDates_ function selects or deselects dates depending on whether they're selected or not (see **line 330** of UserInteractionFunctions.swift).

However in **line 306** it checks if the dates passed to the _selectDates_ function should be selected or not calling the delegate _collectionView(self, shouldSelectItemAt: sectionIndexPath)_.

This means that **it only deselects dates that are marked as "selectable"** by the user. There should be a call to shouldDeselectItemAt if we're deselecting dates ... and the existing check only if we're selecting dates.

That's basically what was causing the weird behaviour in my case.
I managed to get the right behaviour by adding shouldDeselectLogic to my shouldSelect func.

**Comment by reden87** [comment]
@patchthecode sorry for the long unresponsiveness.. here's the whole diff if you start from the latest master branch:

<img width="1221" alt="screen shot 2018-06-04 at 20 01 10" src="https://user-images.githubusercontent.com/869119/40933577-2e9ef5ea-6832-11e8-819e-af5092287ff0.png">

and also I prepared a video about the error:

[bug.zip](https://github.com/patchthecode/JTAppleCalendar/files/2069471/bug.zip)

hope it helps to understand and reproduce the problem.

**Comment by patchthecode** [comment]
ok. finally got around to reviewing this.

@reden87 
I know that you have **programableDeselectCode** in the **didSelectDate** function.
You are correct when you say that with this code, no dates should be deselected.

Your issue is with the indates at the beginning and are somehow still selectable. And the outdates at the end that are still selectable.

Here is the issue:
These dates are out side off the calendar's StartDate and EndDate boundary. Therefore when you call the `deselect()` function, the library says, -> "oh, you are trying to select a date that is outside of the boundary of this calendar". And therefore, nothing happens.

The solution can be 2 thing (i prefer the 2nd solution. Not the first)

1. First we can totally disallow selection of these edge inDates/outDates. But i do not like this solution because it prevents a dev from doing something he may very well want to do.

2. I can leave it up to the developer to manage the selections of indates/outdates himself (which is what is currently happening). 

The real question is: why are you allowing a user to select those edge indates/outdates when it is outside the boundarys of the calendar's legal startDate/endDate?  If you do want them to be selectable, then maybe you should have your startDate to be one month earlier, and your endDate to be one month later.  And if you do not want them to scroll to the whole month, then you can implement a date boundary as i showed in the sample code [here](https://github.com/patchthecode/JTAppleCalendar/issues/365).

Closing this issue, but let me know your thoughts.
